dhall/src/Dhall/Parser/Expression.hs

```diff
@@ -323,13 +323,18 @@ parsers embedded = Parsers {..}
             f <-    (Some <$ _Some <* nonemptyWhitespace)
                 <|> return id
             a <- noted importExpression_
-            b <- Text.Megaparsec.many (try (nonemptyWhitespace *> noted importExpression_))
-            return (foldl app (f a) b)
+            bs <- Text.Megaparsec.many . try $ do
+                (sep, _) <- Text.Megaparsec.match nonemptyWhitespace
+                b <- importExpression_
+                return (sep, b)
+            return (foldl app (f a) bs)
           where
-            app nL@(Note (Src before _ bytesL) _) nR@(Note (Src _ after bytesR) _) =
-                Note (Src before after (bytesL <> bytesR)) (App nL nR)
-            app nL nR =
-                App nL nR
+            app a (sep, b)
+                | Note (Src left _ bytesL) _ <- a
+                , Note (Src _ right bytesR) _ <- b
+                = Note (Src left right (bytesL <> sep <> bytesR)) (App a b)
```

Comment:
I think this would be slightly more readable:

```suggestion
                    = Note (Src left right (bytesL <> sep <> bytesR)) (App a b)
```

Reply:
Not sure I agree ;)

Reply:
What about [Ormolu](https://monadfix.com/ormolu/)-style instead?

```haskell
app a (sep, b)
  | Note (Src left _ bytesL) _ <- a,
    Note (Src _ right bytesR) _ <- b =
    Note (Src left right (bytesL <> sep <> bytesR)) (App a b)
  | otherwise =
    App a b
```

Reply:
That seems like the larger departure from the existing style. Never mind! :)